### PR TITLE
Optimize KMS key reconciliation process

### PR DESCRIPTION
**Pull request by gc-tejas**
This pull request enhances the KMS key reconciliation process by optimizing how missing data keys are detected and handled across multiple scopes and purposes. The main improvement is the introduction of a new repository method that efficiently identifies which scopes lack specific data keys, enabling batch processing and reducing unnecessary lookups. Comprehensive tests have also been added to ensure correctness, including edge cases and large batches. This addresses [ICU-17643](https://hashicorp.atlassian.net/browse/ICU-17643) and [ICU-17641](https://hashicorp.atlassian.net/browse/ICU-17641)

**Key improvements to key reconciliation logic:**

- Added a new repository method, `ListScopesMissingDataKey`, which efficiently returns a map of missing data keys for each purpose across a list of scopes, supporting batching for large requests.
- Refactored the `Kms.ReconcileKeys` method to use the new repository method, allowing it to only process scopes and purposes that are actually missing data keys, improving performance and clarity.

**Testing improvements:**

- Added comprehensive tests for `ListScopesMissingDataKey`, covering cases such as missing keys, missing root keys, all keys present, mixed scenarios, and large batch operations.
- Extended key reconciliation tests to cover scenarios with multiple scopes and mixed missing keys, as well as large batch processing.

**Supporting changes:**

- Added query, `postgresScopesMissingDataKeyQuery` `sqliteScopesMissingDataKeyQuery`, to efficiently find missing data keys for given scopes and purposes.
- Added necessary imports and minor code cleanups to support new functionality and tests.

[ICU-17643]: https://hashicorp.atlassian.net/browse/ICU-17643?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ICU-17641]: https://hashicorp.atlassian.net/browse/ICU-17641?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: extras/kms/repository.go [hunk 1]
```diff
@@ -141,6 +141,68 @@ func (r *repository) list(ctx context.Context, resources interface{}, where stri
...
+	// Process in batches
+	for i := 0; i < len(scopeIds); i += batchSize {
+		end := min(i+batchSize, len(scopeIds))
+		batch := scopeIds[i:end]
+
+		rows, err := reader.Query(ctx, query, []any{batch, purposeStrs})
+		if err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+		defer rows.Close()
```
Comment (YashwanthaGowd) [inline extras/kms/repository.go]: "defer rows.Close()" will only get executed when the function returns. Since defer is used inside a for loop, this will increase the number of open database connections and can cause issues under heavy load. Avoid using "defer" inside the loop.

**Review by ChandraSGadde — DISMISSED** [review]

File: extras/kms/repository.go [hunk 2]
```diff
@@ -141,6 +142,98 @@ func (r *repository) list(ctx context.Context, resources interface{}, where stri
 	return opts.withReader.SearchWhere(ctx, resources, where, args, dbOpts...)
 }
 
+// ListScopesMissingDataKey returns scope IDs from the provided list
+// that are missing a data key for the specified purpose.
+// Supported options:
+//   - WithReader
+func (r *repository) ListScopesMissingDataKey(ctx context.Context, scopeIds []string, purposes []KeyPurpose, opt ...Option) (map[KeyPurpose][]string, error) {
```
Comment (tmessi) [inline extras/kms/repository.go]: Does this need to be exported? If this is only really intended to be used by ReconcileKeys, then it might be good to keep it un-exported instead of adding something new to the public interface of the package.

Comment (gc-tejas) [reply extras/kms/repository.go]: Currently it's only for ReconcileKeys, I will make it unexported.

**Review by ChandraSGadde — APPROVED** [review]
